### PR TITLE
MAILBOX-359 EventFactory should expose builders for events

### PR DESCRIPTION
**Pull request by chibenwa**
Following https://github.com/linagora/james-project/pull/2045#discussion_r242102511 here are builders in EventFactory for Mailbox Events.

This is part of a larger refactoring whose step 2 gonna feature **Dispatcher** removal: Event emitters will directly emit their events using these builders on the DelegatingListener, which will later be the EventBus...

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java [hunk 1]
```diff
@@ -41,7 +41,56 @@
 import org.apache.james.mailbox.model.UpdatedFlags;
 import org.apache.james.mailbox.store.mail.model.Mailbox;
 
+import com.google.common.base.Preconditions;
+
 public class EventFactory {
+    public static class MailboxAddedBuilder {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java]: what about staged builder ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java [hunk 2]
```diff
@@ -56,23 +56,20 @@
...
-    private final EventFactory eventFactory;
 
     @Inject
     public MailboxEventDispatcher(DelegatingMailboxListener delegatingMailboxListener) {
-        this(delegatingMailboxListener, new EventFactory());
+        this((MailboxListener) delegatingMailboxListener);
     }
 
-    private MailboxEventDispatcher(MailboxListener listener, EventFactory eventFactory) {
+    private MailboxEventDispatcher(MailboxListener listener) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java]: you probably don't need this constructor anymore

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageMoveEvent.java [hunk 3]
```diff
@@ -69,11 +70,16 @@ public Builder messageId(Collection<MessageId> messageIds) {
             return this;
         }
 
-        public MessageMoveEvent build() {
+        public Optional<MessageMoveEvent> build() {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageMoveEvent.java]: I don't like the builder taking the decision about building the event or not. I prefer the `isNoop` thing

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageMoveEvent.java]: I will remove this commit.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java]: Staged builder are not good at proposing alrenative build options IMO...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java]: We do still need regisiter "MailboxListener" in tests.

IMO we can add the "delegating" layer in the tests

**Comment by chibenwa** [comment]
Well IMO compusition here will mean:
 - Introducing an intermediate POJO that you could later on use to build your final event like:

```
class MailboxInfo {
   private final User user;
   private final SessionId sessionId;
   private final MailboxId mailboxId;
   private final MailboxPath mailboxPath;
}
```

Not to mention that this complexity is pushed to the caller who will need to do:

```
EventFactory.added()
   .mailboxInfo(MailboxInfo.builder()
        .mailbox(mailbox)
        .session(session)
        .build())
    .metaData(MetaDataInfo.builder()
        .metaData(metaData)
        .build())
    .build();
```

and not

```
EventFactory.added()
        .mailbox(mailbox)
        .session(session)
        .metaData(metaData)
        .build());
```

Which is simpler...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java [hunk 4]
```diff
@@ -34,74 +29,306 @@
...
+        }
+
+        public MailboxDeletionBuilder totalDeletedSize(QuotaSize totalDeletedSize) {
+            this.totalDeletedSize = totalDeletedSize;
+            return this;
+        }
+
+        public MailboxListener.MailboxDeletion build() {
+            mailboxEventChecks();
+            Preconditions.checkState(quotaRoot != null, "Field `quotaRoot` is compulsory");
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java]: `Preconditions.checkNotNull`?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java (hunk 1 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java]: almost all your parameter are mandatory, it makes sense to use staged builder in this context.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java]: Well IMO that's a state stuff: you did not set the state of your builder well which result in something being null.

If you present it like this, the state information becomes more important IMO

**Comment by chibenwa** [comment]
Was :green_apple: 

test this please

**Comment by openpaas-james** [comment]
Build triggered. sha1 is merged.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**